### DISCUSSION TITLE
2D NMR - slices extraction

### DISCUSSION THREAD

**Discussion by ckouvatas** [spectrochempy/spectrochempy#700] [Ideas]
Dear all,
I would like to know if it would be possible to developp a simple way to process 2D NMR spectra recorded on Bruker AVANCE III spectrometers (extraction of slices on the direct or indirect dimension, fitting of MQMAS etc).
Or even extract spectra from pseudo 2D experiments (relaxation time measurements or diffusion experiments)?
Thank you, and have a nice day!
Best regards,
Cassandre K.

**Comment by fernandezc** [comment]
Hi Cassandre, 

Thanks for starting this discussion. In fact, I think most of the tools are already available. The only thing that really needs more work is the processing of FID and SER data (although there are already functional parts).
But if you can work from spectra already processed by Topspin, there shouldn't be any problem.  To read the files, just use the `read_topspin` function to create a dataset - then the slice extration methods and others will be usable.

I'll try to attach a working example as soon as possible (probably tonight). 
If you have specific example to treat, don't hesitate to attach the data or send it to me by email.

Christian
PS; regarding fitting of MQMAS, I can add this functionality!

**Comment by fernandezc** [comment]
Well, in trying to do the example, I've noticed a few problems with the processing of NMR data. It's not a big deal, but needs to be corrected to make the processing fully functional.  So I'll be back here a little later than planned.

**Comment by fernandezc** [comment]
Hi Cassandre,

Few bugs have been corrected (Hopefully!). 

You can find a first example to read, slice and analyse (peak picking, fitting) of a 2D NMR spectra, here: [NMR processing example](https://www.spectrochempy.fr/latest/gettingstarted/examples/gallery/auto_examples_processing/nmr/plot_processing_nmr.html#sphx-glr-gettingstarted-examples-gallery-auto-examples-processing-nmr-plot-processing-nmr-py).

I'll then write an example of how to analyze a series of cross-polarized NMR spectra (probably next week). 

Christian

**Reply by fernandezc** [reply]
Note, for the example to work, you need to update to the last version: 0.6.7.

**Reply by ckouvatas** [reply]
Christian,
Thank you so much ! 
I'm gonna try it today or tomorrow !
I have some interesting data to analyse :)
Have a nice day,
Cassandre

**Reply by ckouvatas** [reply]
Christian, hi again!
It works fine.
But i have some trouble for pseudo-2D data.
Indeed, for instance,  relaxation-time measurements are done with a list ("vdlist" in topspin), and these are the values that are used in the indirect dimension.
However, when we launch the experiment we enter a "dummy value" ("in_0" in topspin), but it is not used afterwards.
It is this "dummy" value that is read by default in spectrochempy (corresponding to "y" of a dataset), and incremented of the same value for each "slice".
How can i do to get the right values (from vdlist)?
Create a list? write a line with a path to reach the "vdlist"?
Thank you, i know this might be trivial.
Cassandre

**Reply by fernandezc** [reply]
Hi Cassandre
I'm currently working on an example with CP Data, and I've found a few problems (in fact SpectroChemPy hasn't yet been tested in depth with NMR, so it's interesting to do so now.

As for the VDList, it's not included at the moment, but it's easy enough to make. Can you send me some data so I can check and write a script for you? Send them by e-mail if you like.

**Reply by fernandezc** [reply]
By the way, if you just need to change the y-coordinates to values from a vdlist (which, if I remember correctly, is simply a text file), you can recreate a coordinate axis from this list:

example:

```python
dataset = scp.read_topspin(
    nmrdir / "test" / "94" / "pdata" / "1" / "2rr"
)

with open(nmrdir / "94" / "vdlist", mode="r") as f:
    vd = [float(val) for val in f.readlines()]
    
dataset.y = scp.Coord(vd, title="time", units="s")
```

To fit the relaxation curve, I can show you how to do if you send me some real data.

**Reply by ckouvatas** [reply]
Christian,
thank you!
I tried yesterday to write somthing that seems to work (and in fact ressemble more or less to the code you proposed, with the changing of y-coordinates), but maybe it is not very elegant!
For pseudo-2D data:

datadir = scp.preferences.datadir

NMR_data = datadir / r"my_nmr_path" 

my_file = scp.read_topspin(my_nmr_path / r"path_to_my_file" / "expno" / "ser")
my_file_processed=scp.fft(my_file)       #fft in the direct dimension, for each spectrum 

lines = loadtxt(r"path_to_vdlist/name_of_vdlist_file")
d1D = NDDataset(lines)

my_file_processed.y=d1D.data

my_file_processed.y.ito("units_wanted",force=True)



I can e-mail you the files !

Cassandre

**Reply by fernandezc** [reply]
It's ok, just change it to `d1D=Coord(...., tile="whatever", units="units_wanted")` as there's no need to create a new NDDataset here, and you can set the units and axe title in one go.

In the next version (still under development), new coordinates will be added for metadata (such as pulse lengths, delays, etc.) that change between experiments (when stored with varying  expnos). 

For pseudo 2D, it would be possible to add an option to define y-coordinates from a list (as present in the expnos folder).

**Reply by fernandezc** [reply]
and of course use `my_file_processed.y=d1D` instead of `my_file_processed.y=d1D.data` to not loose the title and units information.

**Reply by ckouvatas** [reply]
Christian,
It works fine, thank you (and also for the example you made on spectrochempy's website).
Would it be possible to include an option that allows the user to call a different name for the list used for the y coordinate? (it is not alway "vdlist").
Right now the example (pseudo2D data) includes:
" dataset = scp.read_topspin(nmrdir / "relax" / "100" / "ser", use_vdlist=True) "
Thank you again !

**Reply by fernandezc** [reply]
Yes, no problem, I will work on this and this will be available in the next release, hopefully.

Probably, I can change the parameter 
`use_vdlist = False/True`
 to 
`use_list=<name_of_the_list_to_use> or None`

**Reply by fernandezc** [reply]
Cassandre, 

You can now test the latest "dev" version and tell me if it does the job (I did not check with a file different from "vdlist"): 
Now the reader uses "use_list" instead of "use_vdlist" parameter which I had defined before. You can specify the name of the list to read:

```
dataset = scp.read_topspin(nmrdir / "relax" / "100" / "ser", use_list="vdlist")
```

see updated doc:
https://www.spectrochempy.fr/latest/reference/generated/spectrochempy.read_topspin.html#spectrochempy.read_topspin

**Comment by atravert** [comment]
Hi Cassandre and Christian,

I have  made a repository - you should have received the invitation - to share data and scripts related to PFG NMR data (and more if needed :)

A+
Arnaud

**Reply by ckouvatas** [reply]
Thank you Arnaud!
It should be better !
